core/extension/gdextension_interface.h

```diff
@@ -287,6 +287,29 @@ typedef struct {
 	void *class_userdata; // Per-class user data, later accessible in instance bindings.
 } GDExtensionClassCreationInfo;
 
+typedef void (*GDExtensionClassNotification2)(GDExtensionClassInstancePtr p_instance, int32_t p_what, bool p_reversed);
+
+typedef struct {
+	GDExtensionBool is_virtual;
+	GDExtensionBool is_abstract;
+	GDExtensionClassSet set_func;
+	GDExtensionClassGet get_func;
+	GDExtensionClassGetPropertyList get_property_list_func;
+	GDExtensionClassFreePropertyList free_property_list_func;
+	GDExtensionClassPropertyCanRevert property_can_revert_func;
+	GDExtensionClassPropertyGetRevert property_get_revert_func;
+	GDExtensionClassNotification notification_func; // Deprecated - should be nullptr. Only kept for backward compatibility.
```

Comment:
The comment is not really clear -- "should be nullptr", why is it there then? Under which circumstances does an extension developer need to provide this value?

If someone uses `classdb_register_extension_class2`, that sets their minimum Godot version to 4.2; so they cannot support older extensions -- so the compat method wouldn't need to be necessary?

Reply:
I have updated the comment and made it more clear.

If someone uses in Godot 4.2 `classdb_register_extension_class`, they still need `notification_func` to be available, so it is necessary, to keep this available.
Please see https://github.com/godotengine/godot/pull/78634#discussion_r1287495323 for why I added the comment.

Reply:
Thanks 🙂 

> If someone uses in Godot 4.2 `classdb_register_extension_class`, they still need `notification_func` to be available, so it is necessary, to keep this available.

But `classdb_register_extension_class` accepts a different struct: `GDExtensionClassCreationInfo`.

Why `GDExtensionClassCreationInfo2` needs the field `notification_func` field is still not clear to me. That struct can only be used with `classdb_register_extension_class2`, which sets the API minimum version to 4.2. At this point one should _always_ use the `notification_func2`, if I understand correctly.

Maybe it's because of @dsnopek's [comment here](https://github.com/godotengine/godot/pull/78634#discussion_r1287496694) suggesting to reduce code duplication? In that case, I'd argue this is an 
implementation detail that should **not** leak to the user-facing API, because it has no valid use case.

Maybe I'm missing something, but I'd probably remove that field and have a
```cpp
GDExtensionClassNotification2 notification_func;
```
instead, even if it means the C++ code has a bit of duplication.

@Sauermann @dsnopek what do you think?

Reply:
Yeah, we could probably get rid of `notification_func` (or keep the name, but change the type) on the new  `GDExtensionClassCreationInfo2` struct, because it's not really needed in the interface.

That would make the implementation of `_register_extension_class()` a little more complicated, though, as currently it builds the new struct and calls `_register_extension_class2()`. We'd need to either go back to duplicating code in `_register_extention_func()` and `_register_extension_func2()`, or come up with a way to refactor them so they both used a shared 3rd function, or something like that?

Internally, we still need to keep both `notification_func` and `notification_func2` on the `ObjectGDExtension` struct, so I don't think it's the worst thing for them to both be on the interface struct too? I guess it's a question of how much we want to try and hide the internals here.

Reply:
Ack, @Bromeon you and I cross posted :-) Yeah, I think we're both saying the same thing

Reply:
Thanks for confirming! 

I think the Godot C++ implementation (what is stored inside `native_info`) and the GDExtension API will likely diverge _more_ over time, as new features are provided.

With the current approach, that would mean the new API needs to tag along all possible features from previous versions, even though they no longer make sense for an extension developer. Now it's just 1 field, but if we ever have `GDExtensionClassCreationInfo3`, we'd _still_ need to keep around `notification_func` from v1, plus whatever was deprecated in v2.

---

Regarding implementation, this option would be quite nice duplication-wise:
```cpp
struct ClassCreationInfo {
   // All the new stuff
   GDExtensionClassNotification2 v2;
   
   // Legacy features as single fields
   GDExtensionClassNotification notification_func1;
};
```

Reply:
This conversation made me realize we missed a compatibility issue around the `GDExtensionScriptInstanceInfo` struct! I made a new comment about it below.

> I think the Godot C++ implementation (what is stored inside native_info) and the GDExtension API will likely diverge more over time, as new features are provided.
>
> With the current approach, that would mean the new API needs to tag along all possible features from previous versions, even though they no longer make sense for an extension developer. Now it's just 1 field, but if we ever have `GDExtensionClassCreationInfo3`, we'd _still_ need to keep around `notification_func` from v1, plus whatever was deprecated in v2.

Personally, I don't think having a few deprecated fields accruing over time is the worst thing. Maintaining compatibility is always going to lead to some ugliness/messiness somewhere, and it's really just a matter of where you want the ugliness to live.

The current PR sort of shares the ugliness equally with the interface (`gdextension_interface.h`) and the internal Godot implementation. If we prioritize making `gdextension_interface.h` as clean as possible, that means increasing the ugliness in the internal Godot implementation. In particular, I think the steps to do this look something like:

1. Make a new struct based on `GDExtensionScriptInstanceInfo` but only for internal use by `ScriptInstanceExtension`, which has both notification fields. Godot would copy the data from `GDExtensionScriptInstanceInfo` or `GDExtensionScriptInstanceInfo2` for use internally.
2. We don't need to do the same thing for `GDExtensionClassCreationInfo` because we already have `ObjectGDExtension` for use internally
3. `GDExtensionScriptInstanceInfo` and `GDExtensionScriptInstanceInfo2` as well as `GDExtensionClassCreationInfo` and `GDExtensionClassCreationInfo` would each only have one `notification_func` - the old one on the old structs, and the new one on the new structs.
4. Refactor or duplicate the code in `_register_extension_class()`/`_register_extension_class2()` and `gdextension_script_instance_create()`/`gdextension_script_instance_create2()`, to accept the interface structs and copy the data on to the internal structs, rather than just having the older function construct the newer struct and pass it to that function.

This is all totally doable! But it is a bunch of extra code to maintain.

The question is: do we prioritize keeping `gdextension_interface.h` as clean as possible?

Reply:
Thanks for elaborating the precise effort needed in the implementation! You clearly know that part better than me 🙂 



> If we prioritize making `gdextension_interface.h` as clean as possible, that means increasing the ugliness in the internal Godot implementation. [...]
> The question is: do we prioritize keeping `gdextension_interface.h` as clean as possible?

My personal opinion is that I would trade off implementation complexity, if it means that the API can be nicer. A few reasons for that:

* Typically, the number of people coming in contact with the API is much bigger than those with the implementation, so improvements have a greater impact. Also, API users are less familiar with the matter than maintainers.
* APIs are set in stone once published, while implementation can always be refactored. In particular, other requirements (maybe Godot-internal ones, maybe motivated by new GDExtension features) can prompt a code reorganization.
* The GDExtension API in particular has quite a big surface and is rather complex due to its low-level nature. The fewer pitfalls or unusual constraints there are (like "please always provide `NULL` for this pointer"), the easier and less frustrating it becomes to learn it. 

Of course, I have a biased view as I'm mostly a consumer of GDExtension, but maybe _because_ I went through a lot of the early hiccups, I appreciated well-defined semantics 😉 and I think the header has gone great lengths to this already, especially in 4.1! But I definitely don't want to diminish efforts required to maintain the Godot internals; you can probably judge that aspect better than me 🙂 

---



> 1. Make a new struct based on `GDExtensionScriptInstanceInfo` but only for internal use by `ScriptInstanceExtension`, which has both notification fields.

Here, would my previous suggestion work? Then you can always use the latest API struct and only keep the legacy fields separate:
```cpp
struct ClassCreationInfo {
   // All the new stuff
   GDExtensionClassCreationInfo2 latest;
   
   // Legacy features as single fields
   GDExtensionClassNotification legacy_notification_func;
};
```



> 4\. Refactor or duplicate the code in `_register_extension_class()`/`_register_extension_class2()` and `gdextension_script_instance_create()`/`gdextension_script_instance_create2()`, to accept the interface structs and copy the data on to the internal structs (rather than just having the older function construct the newer struct and pass it to that function)

This should also be simpler with the above, and not even need more duplication, just 1 indirection:
```cpp
// Add new internal method, with 1 extra parameter:
void GDExtension::_register_extension_class_internal(
        ...,
        const GDExtensionClassCreationInfo2 *p_extension_funcs,
        GDExtensionClassNotification legacy_notification_func,
) {
	// Do the actual registration
	GDExtension *self = reinterpret_cast<GDExtension *>(p_library);
	...
	extension->gdextension.library = self;
	extension->gdextension.parent_class_name = parent_class_name;
	extension->gdextension.class_name = class_name;
	...
}

// Registration v1
void GDExtension::_register_extension_class(...,  const GDExtensionClassCreationInfo *p_extension_funcs) {
	// First, copy everything, as it is done now. Leave notification_func NULL.
	const GDExtensionClassCreationInfo2 class_info2 = { ... };

	// Delegate to internal, providing a legacy pointer.
	_register_extension_class_inner(..., &class_info2, p_extension_funcs->notification_func);
}

// Registration v2
void GDExtension::_register_extension_class2(...,  const GDExtensionClassCreationInfo2 *p_extension_funcs) {
	// Delegate to internal, providing NULL for the legacy pointer.
	_register_extension_class_inner(..., p_extension_funcs, NULL);
}
```

Reply:
> Here, would my previous suggestion work? Then you can always use the latest API struct and only keep the legacy fields separate

Yes, I think that would work. However, at this point, I think the most important thing would be to get alignment on priorities. I'd really love to hear what other GDExtension team members think about this too! I put a "call for comments" in RocketChat, but this may end up needing to be a topic at a GDExtension meeting.

Reply:
Makes total sense! 👍

Reply:
As mentioned below, at the GDExtension meeting we decided to prioritize keeping `gdextension_interface.h` clean, so I'm going to mark this conversation as resolved